mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,468 @@
+  ): Integer = {
+    tk match {
+      // case _: Token.Ident => // in case of Ident is
+
+      // Alphanumeric keywords
+      case _: Token.ModifierKeyword => getTypeId(SemanticTokenTypes.Modifier)
+      case _: Token.Keyword => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.KwNull => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.KwTrue => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.KwFalse => getTypeId(SemanticTokenTypes.Keyword)
+
+      // extends Symbolic keywords
+      case _: Token.Hash => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Viewbound => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.LeftArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Subtype => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.RightArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Supertype => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.At => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Underscore => getTypeId(SemanticTokenTypes.Variable)
+      case _: Token.TypeLambdaArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.ContextArrow => getTypeId(SemanticTokenTypes.Operator)
+
+      // Constant
+      case _: Token.Constant.Int | _: Token.Constant.Long |
+          _: Token.Constant.Float | _: Token.Constant.Double =>
+        getTypeId(SemanticTokenTypes.Number)
+      case _: Token.Constant.String | _: Token.Constant.Char =>
+        getTypeId(SemanticTokenTypes.String)
+      case _: Token.Constant.Symbol => getTypeId(SemanticTokenTypes.Property)
+
+      // Comment
+      case _: Token.Comment => getTypeId(SemanticTokenTypes.Comment)
+
+      // Interpolation
+      case _: Token.Interpolation.Id | _: Token.Interpolation.SpliceStart =>
+        getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Interpolation.Start | _: Token.Interpolation.Part |
+          _: Token.Interpolation.SpliceEnd | _: Token.Interpolation.End =>
+        getTypeId(SemanticTokenTypes.String) // $ symbol
+
+      // Default
+      case _ => -1
+    }
+
+  }
+  def pickFromTraversed(tk: scala.meta.tokens.Token): Option[NodeInfo] = {
+
+    val adjustForBacktick: Int = {
+      var ret: Int = 0
+      val cName = tk.text.toCharArray()
+      if (cName.size >= 2) {
+        if (
+          cName(0) == 96.toChar // backtick
+          && cName(cName.size - 1) == 96.toChar
+        ) ret = 2
+      }
+      ret
+    }
+
+    val buffer = ListBuffer.empty[NodeInfo]
+    for (node <- nodes) {
+      if (
+        node.pos.start == tk.pos.start &&
+        node.pos.end + adjustForBacktick == tk.pos.end
+      ) buffer.++=(List(node))
+    }
+
+    buffer.toList.headOption
+  }
+
+  case class NodeInfo(
+      sym: Option[Symbol],
+      pos: scala.reflect.api.Position
+  )
+  object NodeInfo {
+    def apply(tree: Tree, pos: scala.reflect.api.Position): NodeInfo =
+      new NodeInfo(Some(tree.symbol), pos)
```

Comment:
```suggestion
      val sym = tree.symbol
      if (sym != NoSymbol && sym != null)
        NodeInfo(Some(tree.symbol), pos)
      else NodeInfo(None, pos)
```

though maybe it's not useful to have a None Symbol? Mayb sym should always be defined otherwise don't add the node?